src/cache/cache.py

```diff
@@ -133,8 +138,11 @@ def pr_write(self, address: int) -> bool:
             return False  # MISS
 
     def cache_line_fill(self, address: int, state: MESIState) -> Optional[CacheLine]:
-        """Fill a cache line with data and set MESI state
-        This would called after a cache miss to fill the cache line with data
+        """
+        Fill a cache line with data and set MESI state. This would be called after a cache miss
+        to fill the cache line with data. The method has run time warnings for invalid operations,
+        to be presented to the user. A cache fill would be followed by a processor read or write, and
+        the MESI state should be set using the set_line_state method.
```

Comment:
I agree it's a bit confusing. I was trying to say that after filling a cache line you'd need to provide the data to the processor in the event of a processor read request, or write the data to the newly filled cache line in the event of a processor write request. There would be no bus operation after a cache fill, only a message to the L1 cache.

Reply:
Sorry if my comment got confusing as to what I'm wondering about for the overall flow, I rewrote this comment a couple of times after I got down to reviewing the tests...

Based on our case analyses, we're going to issue a bus operation on any cache miss, right? Because we need to tell the other caches that we're trying to access data at an address in DRAM, and they might need to update their state if they have that data, or flush the data, etc. And after any miss, we're going to attempt to fill the missed cache line. So I'm wondering what that flow looks like. Here's a partially fleshed-out example:

```python
# from event handler
hit = cache.pr_read(addr)
if (!hit)
    snoop_result = bus_operation(READ, addr)
    # if snoop_result is HITM, might need to back off until data is flushed
    # if snoop_result is HIT or NOHIT, assume other caches updated state based on our bus op
    if (HIT)
        state = SHARED
    else
        state = EXCLUSIVE
    victim_line = cache_fill(addr, state)
    # if victim_line, need to request L1 evict a line
# Regardless of hit or miss, send data to L1, etc.
```

I'm just trying to walk through what all these return values are and whether they give us what we need at each step in order to know how to proceed. I think that they do based on what we know now, and that we can modify if needed in the future, but I wanted to make sure we weren't missing something we can identify now based on having done our case analyses and with knowledge of the state transitions and bus ops.

Reply:
@nkanderson Yeah this looks like the correct flow. One note is that on a cache fill, we will want to issue ` MessageToCache(SENDLINE, Address)` so that L1 will "have" the data as well. We also want to check the victim line in the case of a modified line to do a flush or write back.

Reply:
> One note is that on a cache fill, we will want to issue `MessageToCache(SENDLINE, Address)` so that L1 will "have" the data as well

Good point. This will happen on every request from L1 though, right? Like regardless of whether we had to do a cache fill? So I think as long as the event handler can determine that data returned successfully (or the line was retrieved or whatever), it can handle that.

> We also want to check the victim line in the case of a modified line to do a flush or write back

Good call, and I think you've got this set up to return the `victim_line` or `None`, so the calling code can make that determination, right?